### PR TITLE
[SYCL] Correct handling of enumeration specialization constants

### PR DESCRIPTION
**Pull request by victor-eds**
Lower enumeration specialization constants to `SpecConstant` in SPIR-V instead of `SpecConstantComposite`, thus handling them using their underlying integral types.

### CODE REVIEW

File: sycl/test/basic_tests/SYCL-2020-spec-constants.cpp [hunk 1]
```diff
@@ -4,7 +4,7 @@
 // RUN: sycl-post-link %t.bc -spec-const=emulation -o %t-split2.txt
 // RUN: cat %t-split2_0.prop | FileCheck %s -check-prefixes=CHECK,CHECK-DEF
 // RUN: llvm-spirv -o %t-split1_0.spv -spirv-max-version=1.1 -spirv-ext=+all %t-split1_0.bc
-// RUN: llvm-spirv -o %t-split2_0.spv -spirv-max-version=1.1 -spirv-ext=+all %t-split2_0.bc
+// RUN: llvm-spirv -o - --to-text %t-split1_0.spv | FileCheck %s -check-prefixes=CHECK-SPV
```
Comment (victor-eds) [inline sycl/test/basic_tests/SYCL-2020-spec-constants.cpp:7]: Why were we checking the emulation path for SPIR-V targets, which have native support for specialization constants? I may be missing something, but I don't see the value of such test, as specialization constants would be `gep + load` of a kernel argument. Replace with test of generated SPIR-V for native support.

File: sycl/test/basic_tests/SYCL-2020-spec-constants.cpp [hunk 2]
```diff
@@ -145,3 +151,40 @@ int main() {
...
+// CHECK-SPV-DAG: SpecConstant [[#I32TY]] [[#SPEC4]] 52
+// CHECK-SPV-DAG: SpecConstant [[#I32TY]] [[#SPEC5]] 46
+// CHECK-SPV-DAG: SpecConstant [[#I64TY]] [[#SPEC6]] 27 0
+// CHECK-SPV-DAG: SpecConstant [[#I64TY]] [[#SPEC7]] 81 0
+// CHECK-SPV-DAG: SpecConstant [[#F32TY]] [[#SPEC8]] 1144094720
+// CHECK-SPV-DAG: SpecConstant [[#F64TY]] [[#SPEC9]] 0 1076232192
+// CHECK-SPV-DAG: SpecConstant [[#I32TY]] [[#SPEC10]] 12
+// CHECK-SPV-DAG: SpecConstant [[#I32TY]] [[#SPEC11]] 13
+// CHECK-SPV-DAG: SpecConstant [[#I32TY]] [[#SPEC12]] 2
+// CHECK-SPV-DAG: SpecConstantComposite [[#COMPTY]] {{.*}} [[#SPEC10]] [[#SPEC11]]
```
Comment (victor-eds) [inline sycl/test/basic_tests/SYCL-2020-spec-constants.cpp:190]: New test.

**Comment by victor-eds** [comment]
@AlexeySachkov following on our conversation in #12703. Can you take a look at my proposed solution?

**Comment by maarquitos14** [comment]
@victor-eds the pre-commit failure is known and has been reported already in https://github.com/intel/llvm/pull/12791. I don't think it's related to your changes.

**Review by maarquitos14 — APPROVED** [review]
LGTM.

File: sycl/test/basic_tests/SYCL-2020-spec-constants.cpp (hunk 1 above)
Comment (AlexeySachkov) [reply sycl/test/basic_tests/SYCL-2020-spec-constants.cpp:7]: > Why were we checking the translation to SPIR-V for the emulation path when SPIR-V has native support for this?

It was 3 years ago and I don't remember already :)

The comment below says:

> The test checks that the tool chain correctly identifies all specialization
> constants, emits correct specialization constats map file and can properly
> translate the resulting bitcode to SPIR-V.

I guess my intent was to have some kind of integration test which would also check that whatever we emit for the emulation path can be properly translated into SPIR-V. It is probably not that bad to have it, but at the same time I don't see any special value in it, so I have no objections to remove it.

**Review by AlexeySachkov — APPROVED** [review]

File: sycl/test/basic_tests/SYCL-2020-spec-constants.cpp (hunk 1 above)
Comment (victor-eds) [reply sycl/test/basic_tests/SYCL-2020-spec-constants.cpp:7]: I'd argue we can remove, but I don't have a super strong opinion. Do rest of changes LGTY?

**Comment by sommerlukas** [comment]
Windows CI failure unrelated.

**Comment by sommerlukas** [comment]
Unrelated failure in post-commit reported here: https://github.com/intel/llvm/issues/12956

**Pull request merged**